boa/src/builtins/object/iter.rs

```diff
@@ -0,0 +1,422 @@
+    #[inline]
+    pub fn string_properties(&self) -> StringProperties<'_> {
+        StringProperties(self.properties.iter())
+    }
+
+    /// An iterator visiting all string keys in arbitrary order. The iterator element type is `&'a RcString`.
+    ///
+    /// This iterator does not recurse down the prototype chain.
+    #[inline]
+    pub fn string_property_keys(&self) -> StringPropertyKeys<'_> {
+        StringPropertyKeys(self.properties.keys())
+    }
+
+    /// An iterator visiting all string values in arbitrary order. The iterator element type is `&'a Property`.
+    ///
+    /// This iterator does not recurse down the prototype chain.
+    #[inline]
+    pub fn string_property_values(&self) -> StringPropertyValues<'_> {
+        StringPropertyValues(self.properties.values())
+    }
+}
+
+/// An iterator over the property entries of an `Object`
+#[derive(Debug, Clone)]
+pub struct Iter<'a> {
+    indexed_properties: hash_map::Iter<'a, u32, Property>,
+    string_properties: hash_map::Iter<'a, RcString, Property>,
+    symbol_properties: hash_map::Iter<'a, RcSymbol, Property>,
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = (PropertyKey, &'a Property);
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((key, value)) = self.indexed_properties.next() {
+            Some(((*key).into(), value))
+        } else if let Some((key, value)) = self.string_properties.next() {
+            Some((key.clone().into(), value))
+        } else {
+            let (key, value) = self.symbol_properties.next()?;
+            Some((key.clone().into(), value))
+        }
+    }
+}
+
+impl ExactSizeIterator for Iter<'_> {
+    #[inline]
+    fn len(&self) -> usize {
+        self.indexed_properties.len() + self.string_properties.len() + self.symbol_properties.len()
+    }
+}
+
+impl FusedIterator for Iter<'_> {}
+
+/// An iterator over the keys (`PropertyKey`) of an `Object`.
+#[derive(Debug, Clone)]
+pub struct Keys<'a>(Iter<'a>);
+
+impl<'a> Iterator for Keys<'a> {
+    type Item = PropertyKey;
+    fn next(&mut self) -> Option<Self::Item> {
+        let (key, _) = self.0.next()?;
+        Some(key)
+    }
+}
+
+impl ExactSizeIterator for Keys<'_> {
+    #[inline]
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl FusedIterator for Keys<'_> {}
+
+/// An iterator over the values (`Property`) of an `Object`.
+#[derive(Debug, Clone)]
+pub struct Values<'a>(Iter<'a>);
+
+impl<'a> Iterator for Values<'a> {
+    type Item = &'a Property;
+    fn next(&mut self) -> Option<Self::Item> {
+        let (_, value) = self.0.next()?;
+        Some(value)
+    }
+}
+
+impl ExactSizeIterator for Values<'_> {
+    #[inline]
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl FusedIterator for Values<'_> {}
+
+/// An iterator over the `Symbol` property entries of an `Object`
+#[derive(Debug, Clone)]
+pub struct SymbolProperties<'a>(hash_map::Iter<'a, RcSymbol, Property>);
```

Comment:
Same as before, I think using a private `inner` attribute would be better, so that we don't expose the API we don't want to use.